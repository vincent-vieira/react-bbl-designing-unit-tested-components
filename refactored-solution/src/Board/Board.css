.game-board {
  display: flex;
  flex-direction: column;
}

.board-row {
  display: flex;
  flex-direction: row;
}
