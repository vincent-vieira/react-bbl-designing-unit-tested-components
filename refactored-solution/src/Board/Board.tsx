import React from 'react';
import Square from '../Square/Square';

interface BoardProps {
  size: number;
  squares: string[];
  playSquare: (squareIndex: number) => void;
}

function range(size: number): number[] {
  return [...Array(size).keys()];
}

const Board: React.FC<BoardProps> = ({ size, squares, playSquare }) => {
  return (
    <div role="grid">
      {range(size).map((rowIndex) => (
        <div className="board-row" role="row" key={rowIndex}>
          {range(size).map((cellIndex) => {
            const gridIndex = cellIndex + rowIndex * size;
            const playIndex = gridIndex + 1;

            return (
              <div role="cell" key={cellIndex}>
                <Square
                  playerName={squares[gridIndex]}
                  onClick={() => playSquare(playIndex)}
                />
              </div>
            );
          })}
        </div>
      ))}
    </div>
  );
};

export default Board;
