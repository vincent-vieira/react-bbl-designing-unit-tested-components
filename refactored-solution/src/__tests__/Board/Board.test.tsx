import { render, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Board from '../../Board/Board';

function rangeTo(size: number): number[] {
  return [...Array(size).keys()];
}

describe('Board component', () => {
  describe.each`
    size | expectedElementsCount
    ${3} | ${9}
    ${4} | ${16}
  `('with size $size * $size', ({ size, expectedElementsCount }) => {
    it('should be rendered', async () => {
      render(<Board size={size} squares={[]} onSquareClicked={() => {}} />);

      await waitFor(() => {
        const grid = screen.queryByRole('grid');
        expect(grid).toBeInTheDocument();
        expect(grid).toMatchSnapshot();
      });
    });

    it(`should render ${size} rows`, async () => {
      render(<Board size={size} squares={[]} onSquareClicked={() => {}} />);

      await waitFor(() => {
        expect(screen.queryAllByRole('row')).toHaveLength(size);
      });
    });

    it(`should render ${expectedElementsCount} cells`, async () => {
      render(<Board size={size} squares={[]} onSquareClicked={() => {}} />);

      await waitFor(() => {
        expect(screen.queryAllByRole('cell')).toHaveLength(
          expectedElementsCount
        );
      });
    });

    it.each(
      rangeTo(expectedElementsCount).map((playIndex) => [
        playIndex,
        playIndex + 1,
      ])
    )(
      'should notify parent component when clicking on square number %i',
      async (gridIndex, playIndex) => {
        const playSquareSpy = jest.fn();

        render(
          <Board size={size} squares={[]} onSquareClicked={playSquareSpy} />
        );

        const { findByRole } = within(
          (await screen.findAllByRole('cell'))[gridIndex]
        );

        userEvent.click(await findByRole('button'));

        await waitFor(() => {
          expect(playSquareSpy).toHaveBeenCalledTimes(1);
          expect(playSquareSpy).toHaveBeenNthCalledWith(1, playIndex);
        });
      }
    );
  });
});
