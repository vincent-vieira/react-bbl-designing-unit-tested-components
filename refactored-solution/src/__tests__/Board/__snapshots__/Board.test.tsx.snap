// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component with size 3 * 3 should be rendered 1`] = `
<div
  class="game-board"
  role="grid"
>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
</div>
`;

exports[`Board component with size 4 * 4 should be rendered 1`] = `
<div
  class="game-board"
  role="grid"
>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
  <div
    class="board-row"
    role="row"
  >
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
    <div
      role="cell"
    >
      <button
        class="square"
      />
    </div>
  </div>
</div>
`;
